virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1251 @@
+
+	int len = nh->nlmsg_len - NLMSG_LENGTH (sizeof(*ifa));
+
+	for (attr = IFA_RTA (ifa); RTA_OK (attr, len);
+			attr = RTA_NEXT (attr, len)) {
+		switch(attr->rta_type){
+		case IFA_UNSPEC:
+			break;
+		case IFA_ADDRESS:
+			break;
+		case IFA_LOCAL:
+			ipa->addr = ip_addr_to_string(ifa->ifa_family,
+					RTA_DATA (attr));
+			break;
+		case IFA_LABEL:
+			nif.name = strdup((char *) RTA_DATA (attr));
+			break;
+		case IFA_BROADCAST:
+			break;
+		case IFA_ANYCAST:
+			break;
+		case IFA_CACHEINFO:
+			break;
+		default:
+			break;
+		}
+	}
+
+	nif.ip_addrs = ipa;
+
+	return nif;
+}
+
+/*
+ * Get information from rtmsg.
+ */
+struct route parse_rtmsg(const struct nlmsghdr *nh)
+{
+	struct rtmsg *rtm = (struct rtmsg *) NLMSG_DATA(nh);
+	struct rtattr *attr;
+	int ret;
+	char *dst_ip = NULL;
+
+	struct route rt = {
+		.src = NULL,
+		.dst = NULL,
+		.gw  = NULL,
+		.dev = NULL,
+	};
+
+	int len = RTM_PAYLOAD (nh);
+
+	for (attr = RTM_RTA (rtm); RTA_OK (attr, len);
+			attr = RTA_NEXT (attr, len)) {
+		switch(attr->rta_type){
+		case RTA_UNSPEC:
+			break;
+		case RTA_DST:
+			dst_ip = ip_addr_to_string(rtm->rtm_family,
+					RTA_DATA (attr));
+			ret = asprintf(&(rt.dst), "%s/%d",
+					dst_ip, rtm->rtm_dst_len);
+			if (ret == -1) {
+				printf("Failed allocating memory: %s\n",
+						strerror(errno));
+			}
+			free(dst_ip);
+			break;
+		case RTA_SRC:
+			rt.src = ip_addr_to_string(rtm->rtm_family,
+					RTA_DATA (attr));
+			break;
+		case RTA_IIF:
+			break;
+		case RTA_OIF:
+			rt.dev = iface_idx_to_name(*((int *) RTA_DATA (attr)));
+			break;
+		case RTA_GATEWAY:
+			rt.gw = ip_addr_to_string(rtm->rtm_family,
+					RTA_DATA (attr));
+			break;
+		case RTA_PRIORITY:
+			break;
+		case RTA_PREFSRC:
+			break;
+		case RTA_METRICS:
+			break;
+		case RTA_MULTIPATH:
+			break;
+		case RTA_PROTOINFO:
+			break;
+		case RTA_FLOW:
+			break;
+		case RTA_CACHEINFO:
+			break;
+		default:
+			break;
+		}
+	}
+
+	return rt;
+}
+
+/*
+ * Check if the interface parameters have changed compared to the global list
+ * of interfaces.
+ */
+bool iface_changed(struct iface new, struct iface old)
+{
+	if (new.hw_addr != old.hw_addr ||
+			new.name != old.name ||
+			new.mtu != old.mtu) {
+		return true;
+	}
+
+	return false;
+}
+
+/*
+ * Add a new interface to the list of interfaces.
+ */
+void add_iface_to_list(struct iface nif)
+{
+	iface_list[nif.idx].idx = nif.idx;
+
+	/* Name */
+	if (nif.name) {
+		iface_list[nif.idx].name = strdup(nif.name);
+	}
+
+	/* HW Address */
+	if (nif.hw_addr) {
+		iface_list[nif.idx].hw_addr = strdup(nif.hw_addr);
+	}
+
+	/* MTU */
+	if (nif.mtu) {
+		iface_list[nif.idx].mtu = strdup(nif.mtu);
+	}
+}
+
+/*
+ * Clean an existing interface from the list of interfaces.
+ */
+void delete_iface_from_list(unsigned int idx)
+{
+	free_iface(&(iface_list[idx]));
+
+	struct iface nif = {
+		.idx      = INIT_IF_IDX,
+		.hw_addr  = NULL,
+		.name     = NULL,
+		.mtu      = NULL,
+		.ip_addrs = NULL,
+	};
+
+	iface_list[idx] = nif;
```

Comment:
This looks wrong - you're appending a local variable to a global array?

Reply:
Well, this is not simply to make sure every field of `iface_list[idx]` get reset to the proper values. But I can do something different.